lib/web_git/diff.rb

```diff
@@ -0,0 +1,249 @@
+              start_line = numbers.first.split(" ").first.
+                split(",").first.to_i.abs
+              line_number = start_line + current_line_index
+              file_content += "\n"
+            end
+          end
+          lefts[filenames[i]] = file_content.chomp "\n"
+        end
+        lefts
+      end
+
+      def self.get_each_right(diff)
+        filenames = get_file_names("")
+        files = get_most_of_each_file(diff, filenames)
+        rights = {}
+        files.each_with_index do |file, i|
+          file_content = ""
+          lines = file.split("\n").drop(4)
+          start_line = 0
+          current_line_index = 0
+          line_number = start_line + current_line_index
+          lines.each do |line|
+            # The line numbers in the output of a git diff match this regex
+            # @@ -61,18 +61,15 @@
+            if !line.match?(/@@ ([-]\d+,\d+\s[+]\d+,\d+) @@/)
+              if line.first != "-"
+                file_content += "#{line_number}| " + line + "\n"
+                line_number += 1
+              end
+            else
+              current_line_index = 0
+              numbers = line.scan(/@@ ([-]\d+,\d+\s[+]\d+,\d+) @@/).map(&:join)
+              # If right, start line is the second in a split Array
+              start_line = numbers.first.split(" ").second.
+                split(",").first.to_i.abs
+              line_number = start_line + current_line_index
+              file_content += "\n"
+            end
+          end
+          rights[filenames[i]] = file_content.chomp "\n"
+        end
+        rights
+      end
+
+      def self.get_last_commit_hash
+        Dir.chdir(Rails.root) do
+          log = `git log -1 --oneline`
+          log.split.first
+        end
+      end
+
+      def self.get_most_of_each_file(diff, filenames)
+        files = []
+        lines = diff.split("\n")
+        last_index = 0
+        filenames.count.times do |i|
+          file = filenames[i]
+          content = ""
+          last_index.upto(lines.count - 1) do |index|
+            line = lines[index]
+            if !match_other_files(line, file, filenames)
+              content += line + "\n"
+            else
+              last_index = index
+              break
+            end
+          end
+          files.push content
+        end
+        files
+      end
+
+      def self.match_other_files(line, file, filenames)
+        filenames.each do |other_file|
+          if file != other_file
+            # It looks like:
+            #   --- a/<path-to-file>
+            #   +++ b/<path-to-file>
+            if line.include?('diff --git a/' + other_file + ' b/' + other_file)
+              return true
+            end
+          end
+        end
+        false
+      end
+
+      def self.get_file_names(commit)
+        Dir.chdir(Rails.root) do
+          filenames = `git diff-tree --no-commit-id --name-only -r #{commit}`
```

Comment:
I might consider putting `` filenames = `git diff-tree --no-commit-id --name-only -r #{commit}` `` into an `else` block of the subsequent conditional. That's effectively how it's acting, right?

Reply:
True!